### ISSUE TITLE
GPU training leads to overfits to sub-buckets

### ISSUE THREAD

**Issue by howardchanth** [facebookresearch/PyTorch-BigGraph#225]
Thanks a lot for sharing such useful graph embedding tools! However, when we are training the example data (Freebase 15K) using our GPU and CPU devices, we found that there is an increasing gap between evaluation loss before and after GPU training, which indicates that the GPU training is overfitting. The only possible causal difference that we found between GPU and CPU training is that GPU training further divides edges into sub-buckets, and the sub-buckets are not being shuffled over the epochs. So probably the model will be overfitted to each of the sub-buckets, and the embeddings will be possibly separated into different embedding spaces during training. 

Similar phenomenons also observed when we are producing our own movie recommendation system using PBG. The gap was also observed in GPU training and converged to a stable level (around 10). I am wondering if there is any way that we can solve the potential overfitting problem. Please see below some additional information for your reference and feel free to let me know if there is any additional information needed. Thanks a lot for your time and help!

## Steps to reproduce

  1.  Training Freebase15k on GPU and CPU using respective example config files
  2.  `eval_frac` set to 0.01 instead of the default 0
  3.  `num_gpus` set to 2 for GPU training

## Observed Results

  * There is an increasing gap between the evaluation loss before and after training for the GPU training
  * The gap for CPU is converging to zero as expected
  * Please see below the plots of metrics for your reference
  - ![image](https://user-images.githubusercontent.com/55630770/127325061-1e3a0d96-c693-41f4-a338-b3321fdad36c.png)
  - ![image](https://user-images.githubusercontent.com/55630770/127325083-fcd672f6-670c-418c-8046-711426cd25d7.png)
  - ![image](https://user-images.githubusercontent.com/55630770/127325100-f42ecc67-3927-4b66-a8c0-6d442f5d811e.png)

## Expected Results

  * GPU training on example Freebase 15K data should not have a gap between evaluation loss before and after training (i.e. The gap should converge to zero over epochs)

**Comment by howardchanth** [comment]
Sorry for bringing more trouble but another phenomenon observed is that the evaluation losses (both before and after) for CPU training are increasing over epochs. I am wondering if it is reasonable as the evaluation should be decreasing overtime if the model is not overfitted to training data. Any insights on this problem would be very helpful. Thanks a lot:)

**Comment by adamlerer** [comment]
Hi, thanks for reporting these issues. Regarding the evaluation losses increasing, that *is* the result of overfitting as expected - if you're seeing it you should stop training at that point, or regularize more heavily (smaller dimension or higher weight decay).

The GPU behavior I'm not sure about, let me take a look and get back to you.

**Comment by lw** [comment]
Just one quick thing: the GPU training does indeed sub-partition the entities and thus cause poorer mixing (because it samples negatives for an entity only from its sub-partition rather than from the whole partition), but unless I'm misremembering it should re-partition in a different way during each epoch, so there should not be any long-term correlation due to the sub-partitioning. I'm spitballing here but could this in fact act as a kind of regularization where, when we're already overfitting, each epoch overfits to its sub-partitioning and then the following epoch "breaks" that overfitting by re-overfitting to its new sub-partitioning. I'm not sure whether this is enough to explain the loss "spread" phenomenon.

**Comment by howardchanth** [comment]
Thanks @adamlerer for the insights on increasing evaluation loss! We will give them a shot.

As for the overfitting problem to sub-buckets, I have checked the debugging log during GPU training and the relevant codes and, correct me if I am wrong, the division of subbuckets and the edges seems to be the same over the epochs. I haved added the following print statements to check the permutations of training edges. Please find attached the training log file, as you can see, every epoch the sub-buckets are having the same permutation of the edges (see the values of `edges_lhs`, `edges_rhs`, `edges_rel`) and the subbuckets are basically having the same permutation according to their entities.

[fb15k_debug.txt](https://github.com/facebookresearch/PyTorch-BigGraph/files/6905702/fb15k_debug.txt)

2021-07-30 13:53:33,663   [Trainer-0] Starting epoch 49 / 50, edge path 1 / 1, edge chunk 1 / 1
...
**2021-07-30 13:53:33,820   [Trainer-0] ( 0 , 0 ): edges lhs: tensor([ 6442, 12700,  4870,  ...,  4704,  2257, 13784])
2021-07-30 13:53:33,820   [Trainer-0] ( 0 , 0 ): edges rhs: tensor([ 1261,  1275,  1349,  ...,  8032, 14179,  9205])
2021-07-30 13:53:33,820   [Trainer-0] ( 0 , 0 ): edges rel: tensor([260, 968,  73,  ..., 166, 284, 968])**
...
**2021-07-30 13:53:33,835   [Trainer-0] ( 0 , 0 ): {('all', 0, 0): slice(0, 3738, None), ('all', 0, 1): slice(3738, 7476, None), ('all', 0, 2): slice(7476, 11214, None), ('all', 0, 3): slice(11214, 14951, None)}**

**2021-07-30 13:53:36,928   [Trainer-0] Starting epoch 50 / 50, edge path 1 / 1, edge chunk 1 / 1
...
**2021-07-30 13:53:37,093   [Trainer-0] ( 0 , 0 ): edges lhs: tensor([ 6442, 12700,  4870,  ...,  4704,  2257, 13784])
2021-07-30 13:53:37,093   [Trainer-0] ( 0 , 0 ): edges rhs: tensor([ 1261,  1275,  1349,  ...,  8032, 14179,  9205])
2021-07-30 13:53:37,093   [Trainer-0] ( 0 , 0 ): edges rel: tensor([260, 968,  73,  ..., 166, 284, 968])**
...

**2021-07-30 13:53:37,107   [Trainer-0] ( 0 , 0 ): {('all', 0, 0): slice(0, 3738, None), ('all', 0, 1): slice(3738, 7476, None), ('all', 0, 2): slice(7476, 11214, None), ('all', 0, 3): slice(11214, 14951, None)}**

 I am wondering if setting the seeds in the generator leading to trouble, the `manual_seed` as I understand is applied to the whole CPU and is independent of epoches leading to same random states of every epoch when dividing the buckets :

```
                if num_eval_edges > 0:
                    g = torch.Generator()
                    g.manual_seed(
                        hash((edge_path_idx, edge_chunk_idx, cur_b.lhs, cur_b.rhs))
                    )
                    eval_edge_idxs = torch.randint(
                        num_edges, (num_eval_edges,), dtype=torch.long, generator=g
                    )

```

Similar situations also happen when we are applying to our own user-movie data with multiple entities and relations. The bucket division seems to be invariant to epoches. 
**epoch 1**
2021-07-17 20:13:20,462   [Trainer-0] ( 0 , 0 ): {('director', 0, 0): slice(0, 6164, None), ('director', 0, 1): slice(6164, 12328, None), ('director', 0, 2): slice(12328, 18492, None), ('director', 0, 3): slice(18492, 24653, None), ('country', 0, 0): slice(0, 24, None), ('country', 0, 1): slice(24, 48, None), ('country', 0, 2): slice(48, 72, None), ('country', 0, 3): slice(72, 94, None), ('tag', 0, 0): slice(0, 12214, None), ('tag', 0, 1): slice(12214, 24428, None), ('tag', 0, 2): slice(24428, 36642, None), ('tag', 0, 3): slice(36642, 48854, None), ('userid', 0, 0): slice(0, 3594771, None), ('userid', 0, 1): slice(3594771, 7189542, None), ('userid', 0, 2): slice(7189542, 10784313, None), ('userid', 0, 3): slice(10784313, 14379083, None), ('actor', 0, 0): slice(0, 16368, None), ('actor', 0, 1): slice(16368, 32736, None), ('actor', 0, 2): slice(32736, 49104, None), ('actor', 0, 3): slice(49104, 65472, None), ('language', 0, 0): slice(0, 21, None), ('language', 0, 1): slice(21, 42, None), ('language', 0, 2): slice(42, 63, None), ('language', 0, 3): slice(63, 82, None), ('movieid', 0, 0): slice(0, 37373, None), ('movieid', 0, 1): slice(37373, 74746, None), ('movieid', 0, 2): slice(74746, 112119, None), ('movieid', 0, 3): slice(112119, 149491, None)}
**epoch 2**
2021-07-17 20:53:46,252   [Trainer-0] ( 0 , 0 ): {('director', 0, 0): slice(0, 6164, None), ('director', 0, 1): slice(6164, 12328, None), ('director', 0, 2): slice(12328, 18492, None), ('director', 0, 3): slice(18492, 24653, None), ('country', 0, 0): slice(0, 24, None), ('country', 0, 1): slice(24, 48, None), ('country', 0, 2): slice(48, 72, None), ('country', 0, 3): slice(72, 94, None), ('tag', 0, 0): slice(0, 12214, None), ('tag', 0, 1): slice(12214, 24428, None), ('tag', 0, 2): slice(24428, 36642, None), ('tag', 0, 3): slice(36642, 48854, None), ('userid', 0, 0): slice(0, 3594771, None), ('userid', 0, 1): slice(3594771, 7189542, None), ('userid', 0, 2): slice(7189542, 10784313, None), ('userid', 0, 3): slice(10784313, 14379083, None), ('actor', 0, 0): slice(0, 16368, None), ('actor', 0, 1): slice(16368, 32736, None), ('actor', 0, 2): slice(32736, 49104, None), ('actor', 0, 3): slice(49104, 65472, None), ('language', 0, 0): slice(0, 21, None), ('language', 0, 1): slice(21, 42, None), ('language', 0, 2): slice(42, 63, None), ('language', 0, 3): slice(63, 82, None), ('movieid', 0, 0): slice(0, 37373, None), ('movieid', 0, 1): slice(37373, 74746, None), ('movieid', 0, 2): slice(74746, 112119, None), ('movieid', 0, 3): slice(112119, 149491, None)}

Do you have any insights that which part of code may be investigated to figure out the problem? Thanks a lot again for your time and help!

**Comment by adamlerer** [comment]
Hi Howard, that's a good hypothesis but I don't think this is exactly the issue. The code you're pointing to is what selects the *validation set*, which we *do* want to be constant across epochs. The code that assigns entities to sub-buckets each epoch is here

https://github.com/facebookresearch/PyTorch-BigGraph/blob/e63485a7216703cb545ea37d8ceb646b79306edb/torchbiggraph/train_gpu.py#L271

and I believe it is different each epoch.

**Comment by adamlerer** [comment]
Oh, I have a hypothesis about what's going on... We do this hack to speed up removal of the eval edges in `train_gpu.py` by overwriting them *in place* with train edges (https://github.com/facebookresearch/PyTorch-BigGraph/blob/e63485a7216703cb545ea37d8ceb646b79306edb/torchbiggraph/train_gpu.py#L498): 
```
       if eval_edge_idxs is not None:
            bucket_logger.debug("Removing eval edges")
            tk.start("remove_eval")
            num_eval_edges = len(eval_edge_idxs)
            edges_lhs[eval_edge_idxs] = edges_lhs[-num_eval_edges:].clone()
            edges_rhs[eval_edge_idxs] = edges_rhs[-num_eval_edges:].clone()
            edges_rel[eval_edge_idxs] = edges_rel[-num_eval_edges:].clone()
            edges_lhs = edges_lhs[:-num_eval_edges]
            edges_rhs = edges_rhs[:-num_eval_edges]
            edges_rel = edges_rel[:-num_eval_edges]
            bucket_logger.debug(
                f"Time spent removing eval edges: {tk.stop('remove_eval'):.4f} s"
            )
```

But doesn't this mean that when we do eval after training it's now operating on train edges instead of eval edges? (https://github.com/facebookresearch/PyTorch-BigGraph/blob/e63485a7216703cb545ea37d8ceb646b79306edb/torchbiggraph/train_cpu.py#L674)

@lw PTAL

**Comment by lw** [comment]
Yeah at first sight I agree with that diagnosis. @howardchanth I tried a fix in https://github.com/facebookresearch/PyTorch-BigGraph/pull/231, could you manually patch it in, re-run the GPU training and see if it fixes the discrepancy?
